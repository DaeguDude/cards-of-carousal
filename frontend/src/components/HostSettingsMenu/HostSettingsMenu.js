import React, { useState } from 'react';
import styled from 'styled-components';

import Header from '../Header/Header';

const SettingsMenu = styled.div`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  width: 250px;
  background-color: var(--primary-background-color);
  display: flex;
  flex-direction: column;

  .host-settings-header {
    display: flex;
    align-items: flex-end;
    height: 50px;
  }

  .host-settings-header h3 {
    font-size: 1.5rem;
    line-height: 0.9rem;
    color: var(--secondary-text-color);
  }

  @media (min-width: 800px) {
    width: 400px;

    .host-settings-header {
      height: 130px;
    }

    .host-settings-header h3 {
      font-size: 3.5rem;
      line-height: 2.2rem;
    }
  }

  @media (min-width: 2000px) {
    width: 600px;

    .host-settings-header h3 {
      font-size: 5rem;
      line-height: 3rem;
    }
  }

  @media (min-width: 3500px) {
    width: 1100px;

    .host-settings-header {
      height: 200px;
    }

    .host-settings-header h3 {
      font-size: 8rem;
      line-height: 5rem;
    }
  }
`;

function useSettingsMenuHandler(components) { 
  const initialState = components.map(() => 'enabled');
  const [accordionSettings, setAccordionSettings] = useState(initialState);

  const anyAreOpen = accordionSettings.some(
    (setting) => setting.state === 'open',
  );

  function resetAccordions() {
    setAccordionSettings(initialState);
  }

  function handleAccordionClick(settingIndex) {
    const newSettings = accordionSettings.map((setting, index) => index === settingIndex ? 'open' : 'disabled');
    setAccordionSettings(newSettings);
  }

  function makePopulatedAccordion(Accordion, index) {
    return (
      <Accordion
        key={Accordion.name}
        accordionState={accordionSettings[index]}
        onClickActions={{
          open: resetAccordions,
          enabled: () => handleAccordionClick(index),
          disabled: resetAccordions
        }}
      />
    )
  }

  function makePopulatedButton(Button) {
    return <Button
      key={Button.name}
      isEnabled={!anyAreOpen}
      onDisabledClick={resetAccordions}
    />
  }

  const hydratedComponents = components.map((Component, index) => {
    switch(Component.name) {
      case 'PlayerKicker':
        return makePopulatedAccordion(Component, index);

      case 'JoinCodeHider':
        return makePopulatedButton(Component);

      default: 
        throw new Error(`Please add ${Component.name} to HostSettingsMenu.js in the useSettingsMenuHandler`)
    }
  });

  return [
    hydratedComponents,
    resetAccordions
  ]
}

function HostSettingsMenu({settingsComponentList}) {
  const [components, resetAccordions] = useSettingsMenuHandler(settingsComponentList)

  function onOutsideClick(event) {
    if (event.currentTarget === event.target) {
      resetAccordions();
    }
  }

  return (
    <SettingsMenu onClick={onOutsideClick}>
      <Header className="host-settings-header" onClick={resetAccordions}>
        <h3>SETTINGS</h3>
      </Header>

      {components}
    </SettingsMenu>
  );
}

export default HostSettingsMenu;
