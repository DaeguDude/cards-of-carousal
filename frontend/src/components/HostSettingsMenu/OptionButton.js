import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

const propTypes = {
  isActive: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
  children: PropTypes.string.isRequired,
};

const StyledOptionButton = styled.button`
  color: #dddddd;

  ${(props) =>
    props.isActive &&
    css`
      color: var(--primary-text-color);
      :hover {
        background-color: var(--accent-background-color);
      }
    `}

  background-color: var(--primary-background-color);
  display: flex;
  flex-direction: column;
  border: none;
  padding: 30px 0 20px 25px;

  text-align: left;
  font-size: 1rem;
  font-weight: 700;

  @media (min-width: 800px) {
    font-size: 1.5rem;
  }

  @media (min-width: 2000px) {
    font-size: 2.5rem;
  }

  @media (min-width: 3500px) {
    font-size: 4.5rem;
  }
`;

function OptionButton({
  isActive: optionButtonIsActive,
  onClick: optionClick,
  children,
}) {
  return (
    <StyledOptionButton
      isActive={optionButtonIsActive}
      onClick={optionButtonIsActive ? optionClick : undefined}
    >
      {children}
    </StyledOptionButton>
  );
}

OptionButton.propTypes = propTypes;

export default OptionButton;
