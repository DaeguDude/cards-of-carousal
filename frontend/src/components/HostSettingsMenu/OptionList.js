import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

const propTypes = {
  listContent: PropTypes.arrayOf(PropTypes.string).isRequired,
  isActive: PropTypes.bool.isRequired,
  onAccordionClick: PropTypes.func.isRequired,
  onListItemClick: PropTypes.func.isRequired,
  activeText: PropTypes.string.isRequired,
  inactiveText: PropTypes.string.isRequired,
};

const OptionListButton = styled.button`
  color: var(--primary-text-color);
  background-color: var(--primary-background-color);

  ${(props) =>
    props.isActive &&
    css`
      color: var(--secondary-text-color);
      background-color: var(--secondary-background-color);
    `}

  display: flex;
  flex-direction: column;
  border: none;
  padding: 20px 25px;
  width: 100%;

  text-align: left;
  font-size: 1.5rem;
  font-weight: 700;
`;

const OptionListItemButton = styled(OptionListButton)`
  font-size: 1.3rem;
  padding: 10px 50px;
`;

function OptionList({
  listContent,
  isActive: optionListIsActive,
  onAccordionClick: onOptionListClick,
  onListItemClick,
  activeText,
  inactiveText,
}) {
  return (
    <div>
      <OptionListButton
        type="button"
        onClick={onOptionListClick}
        isActive={optionListIsActive}
      >
        {optionListIsActive ? activeText : inactiveText}
      </OptionListButton>
      {optionListIsActive &&
        listContent.map((listItem) => (
          <OptionListItemButton
            type="button"
            key={listItem}
            onClick={(event) => onListItemClick(event, listItem)}
          >
            {listItem}
          </OptionListItemButton>
        ))}
    </div>
  );
}

OptionList.propTypes = propTypes;

export default OptionList;
