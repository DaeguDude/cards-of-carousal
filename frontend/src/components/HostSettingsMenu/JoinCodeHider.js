import React from 'react';
import PropTypes from 'prop-types';

import OptionButton from './OptionButton';

const propTypes = {
  isEnabled: PropTypes.bool.isRequired,
  onDisabledClick: PropTypes.func.isRequired,
};

// placeholder function
function hideJoinCode() {
  // eslint-disable-next-line
  console.log('PRETEND THE JOIN CODE IS HIDDEN');
}

function JoinCodeHider({ isEnabled, onDisabledClick }) {
  return (
    <OptionButton
      isEnabled={isEnabled}
      onEnabledClick={hideJoinCode}
      onDisabledClick={onDisabledClick}
    >
      HIDE JOIN CODE
    </OptionButton>
  );
}

JoinCodeHider.propTypes = propTypes;

export default JoinCodeHider;
