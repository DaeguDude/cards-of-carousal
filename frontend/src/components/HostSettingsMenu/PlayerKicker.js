import React, { useContext } from 'react';
import PropTypes from 'prop-types';

import OptionList from './OptionList';
import { HostContext } from '../../contexts/HostContext/HostContext';

const propTypes = {
  accordionState: PropTypes.string.isRequired,
  onEnabledClick: PropTypes.func.isRequired,
  onNotEnabledClick: PropTypes.func.isRequired,
};

function PlayerKicker({ accordionState, onEnabledClick, onNotEnabledClick }) {
  const {
    state: { playerIDs, players },
    dispatch,
  } = useContext(HostContext);

  const playerList = playerIDs
    .map((playerId) => players[playerId].name)
    .sort((a, b) => b > a);

  function kickPlayer(playerName) {
    const targetPlayer = playerIDs.find(
      (playerId) => players[playerId].name === playerName,
    );

    dispatch({
      type: 'PLAYER_DISCONNECTED',
      payload: { playerId: targetPlayer },
    });
  }

  return (
    <OptionList
      listContent={playerList}
      state={accordionState}
      onEnabledOptionListClick={onEnabledClick}
      onNotEnabledOptionListClick={onNotEnabledClick}
      onListItemClick={kickPlayer}
      openText="KICK WHO?"
      closedText="KICK PLAYER"
    />
  );
}

PlayerKicker.propTypes = propTypes;

export default PlayerKicker;
