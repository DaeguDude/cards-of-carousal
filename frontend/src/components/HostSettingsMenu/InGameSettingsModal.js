import React from 'react';

import HostSettingsMenu from './HostSettingsMenu';
import PlayerKicker from './SettingsSubComponents/PlayerKicker';
import JoinCodeHider from './SettingsSubComponents/JoinCodeHider';
import JoinCodeShuffler from './SettingsSubComponents/JoinCodeShuffler';

export default function PreGameSettingsModal() {
  return (
    <HostSettingsMenu
      settingsComponentList={[
        { type: 'button', component: JoinCodeHider },
        { type: 'accordion', component: PlayerKicker },
        { type: 'button', component: JoinCodeShuffler },
      ]}
    />
  );
}
