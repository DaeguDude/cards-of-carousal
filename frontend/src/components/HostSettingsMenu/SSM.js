import React from 'react';
import HostSettingsMenu from './HostSettingsMenu';
import PlayerKicker from './PlayerKicker';
import JoinCodeHider from './JoinCodeHider';

function SSM() {
  return (
    <HostSettingsMenu 
      settingsComponentList={[
        PlayerKicker,
        JoinCodeHider
      ]}
    />
  );
}

export default SSM;
