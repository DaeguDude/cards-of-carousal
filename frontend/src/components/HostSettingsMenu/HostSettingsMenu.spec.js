import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

import HostSettingsMenu from './HostSettingsMenu';
import HostProvider from '../../contexts/HostContext/HostContext';

describe('HostSettingsMenu', () => {
  describe('rendering', () => {
    it('renders the settings heading', () => {
      render(
        <HostProvider>
          <HostSettingsMenu />
        </HostProvider>,
      );

      expect(screen.getByText('SETTINGS')).toBeInTheDocument();
    });

    it('renders the hide join code button', () => {
      render(
        <HostProvider>
          <HostSettingsMenu />
        </HostProvider>,
      );

      expect(
        screen.getByRole('button', { name: 'HIDE JOIN CODE' }),
      ).toBeInTheDocument();
    });

    it('renders the kick player button', () => {
      render(
        <HostProvider>
          <HostSettingsMenu />
        </HostProvider>,
      );

      expect(
        screen.getByRole('button', { name: 'KICK PLAYER' }),
      ).toBeInTheDocument();
    });

    it('renders the shuffle join code button', () => {
      render(
        <HostProvider>
          <HostSettingsMenu />
        </HostProvider>,
      );

      expect(
        screen.getByRole('button', { name: 'SHUFFLE JOIN CODE' }),
      ).toBeInTheDocument();
    });

    it('matches the expected snapshot', () => {
      const tree = renderer
        .create(
          <HostProvider>
            <HostSettingsMenu />
          </HostProvider>,
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('functionality', () => {
    // temporary test pending overhall in next sprint
    it('Closes the OptionList when a click is outside of OptionList', () => {
      render(
        <HostProvider>
          <HostSettingsMenu />
        </HostProvider>,
      );

      // open the option list and verify that it is open
      userEvent.click(screen.getByRole('button', { name: 'KICK PLAYER' }));
      expect(screen.getByText('KICK WHO?')).toBeInTheDocument();

      // click on the settings header and verify that the option list is now closed
      userEvent.click(screen.getByText('SETTINGS'));
      expect(screen.queryByText('KICK WHO?')).not.toBeInTheDocument();
    });

    // temporary test pending overhall in next sprint
    it('makes clicking another button not work when the option list is open', () => {
      const consoleSpy = jest
        .spyOn(console, 'log')
        .mockImplementation(() => {});

      render(
        <HostProvider>
          <HostSettingsMenu />
        </HostProvider>,
      );

      // click the skip button once and verify that it logs to the console
      userEvent.click(screen.getByRole('button', { name: 'HIDE JOIN CODE' }));
      expect(consoleSpy).toHaveBeenCalledTimes(1);

      // open the OptionList
      userEvent.click(screen.getByRole('button', { name: 'KICK PLAYER' }));
      expect(screen.getByText('KICK WHO?')).toBeInTheDocument();

      // click the skip button again and verify that it doesn't log to the console this time
      userEvent.click(screen.getByRole('button', { name: 'HIDE JOIN CODE' }));
      expect(consoleSpy).toHaveBeenCalledTimes(1);
    });
  });
});
