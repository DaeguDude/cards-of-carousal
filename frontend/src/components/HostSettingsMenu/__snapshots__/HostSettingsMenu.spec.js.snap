// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HostSettingsMenu rendering matches the expected snapshot 1`] = `
.c1 {
  width: 100%;
  background-color: var(--secondary-background-color);
  color: var(--secondary-text-color);
}

.c3 {
  color: var(--primary-text-color);
  background-color: var(--primary-background-color);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border: none;
  padding: 20px 25px;
  width: 100%;
  text-align: left;
  font-size: 1rem;
  font-weight: 700;
}

.c3:hover {
  background-color: var(--accent-background-color);
}

.c2 {
  color: #dddddd;
  color: var(--primary-text-color);
  background-color: var(--primary-background-color);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border: none;
  padding: 30px 0 20px 25px;
  text-align: left;
  font-size: 1rem;
  font-weight: 700;
}

.c2:hover {
  background-color: var(--accent-background-color);
}

.c0 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  width: 250px;
  background-color: var(--primary-background-color);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 .host-settings-header {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  height: 50px;
}

.c0 .host-settings-header h3 {
  font-size: 1.5rem;
  line-height: 0.9rem;
  color: var(--secondary-text-color);
}

@media (min-width:800px) {
  .c3 {
    font-size: 1.5rem;
  }
}

@media (min-width:2000px) {
  .c3 {
    font-size: 2.5rem;
  }
}

@media (min-width:3500px) {
  .c3 {
    font-size: 4.5rem;
  }
}

@media (min-width:800px) {
  .c2 {
    font-size: 1.5rem;
  }
}

@media (min-width:2000px) {
  .c2 {
    font-size: 2.5rem;
  }
}

@media (min-width:3500px) {
  .c2 {
    font-size: 4.5rem;
  }
}

@media (min-width:800px) {
  .c0 {
    width: 400px;
  }

  .c0 .host-settings-header {
    height: 130px;
  }

  .c0 .host-settings-header h3 {
    font-size: 3.5rem;
    line-height: 2.2rem;
  }
}

@media (min-width:2000px) {
  .c0 {
    width: 600px;
  }

  .c0 .host-settings-header h3 {
    font-size: 5rem;
    line-height: 3rem;
  }
}

@media (min-width:3500px) {
  .c0 {
    width: 1100px;
  }

  .c0 .host-settings-header {
    height: 200px;
  }

  .c0 .host-settings-header h3 {
    font-size: 8rem;
    line-height: 5rem;
  }
}

<div
  className="c0"
  onClick={[Function]}
>
  <header
    className="c1 host-settings-header"
    onClick={[Function]}
  >
    <h3>
      SETTINGS
    </h3>
  </header>
  <button
    className="c2"
    onClick={[Function]}
  >
    HIDE JOIN CODE
  </button>
  <div>
    <button
      className="c3"
      onClick={[Function]}
      type="button"
    >
      KICK PLAYER
    </button>
  </div>
  <button
    className="c2"
    onClick={[Function]}
  >
    SHUFFLE JOIN CODE
  </button>
</div>
`;
